File: src/main/java/linq/Select.java
package linq;

import linq.collections.enumerable.ArrayEnumerable;
import linq.collections.enumerable.EmptyEnumerable;
import linq.exception.ExceptionArgument;
import linq.exception.ThrowHelper;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;

public class Select {
    static <TSource, TResult> Enumerable<TResult> select(Enumerable<TSource> source, Function<TSource, TResult> selector) {
        if (source == null) {
            ThrowHelper.throwArgumentNullException(ExceptionArgument.source);
        }
        if (selector == null) {
            ThrowHelper.throwArgumentNullException(ExceptionArgument.selector);
        }

        if (source instanceof AbstractIterator<TSource> iterator) {
            return iterator.select(selector);
        }

        if (source instanceof ArrayEnumerable<TSource> array) {
            if (array.count() == 0) return new EmptyEnumerable<>();
            return new ArraySelectIterator<>(array, selector);
        }

        return new EnumerableSelectIterator<>(source, selector);
    }

    static <TSource, TResult> Enumerable<TResult> select(Enumerable<TSource> source, BiFunction<TSource, Integer, TResult> selector) {
        if (source == null) {
            ThrowHelper.throwArgumentNullException(ExceptionArgument.source);
        }

        if (selector == null) {
            ThrowHelper.throwArgumentNullException(ExceptionArgument.selector);
        }

        if (Enumerable.isEmptyArray(source)) {
            return Enumerable.empty();
        }

        return new Enumerable<>() {
            @Override
            public Enumerator<TResult> enumerator() {
                return new EnumerableIndexSelectIterator<>(source.enumerator(), selector);
            }
        };
    }

    static final class ArraySelectIterator<TSource, TResult> extends AbstractIterator<TResult> {
        private final ArrayEnumerable<TSource> source;
        private final Function<TSource, TResult> selector;

        public ArraySelectIterator(ArrayEnumerable<TSource> source, Function<TSource, TResult> selector) {
            assert (source != null);
            assert (selector != null);
            assert (source.count() > 0);
            this.source = source;
            this.selector = selector;
        }

        @Override
        protected AbstractIterator<TResult> getClone() {
            return new ArraySelectIterator<>(source, selector);
        }

        @Override
        public boolean moveNext() {
            var source = this.source;
            var index = state - 1;
            if (index < source.count()) {
                state++;
                current = selector.apply(source.get(index));
                return true;
            }

            close();
            return false;
        }

        @Override
        public <TResult2> Enumerable<TResult2> select(Function<TResult, TResult2> selector) {
            return new ArraySelectIterator<>(source, this.selector.andThen(selector));
        }

        @Override
        public List<TResult> toList() {
            var count = source.count();
            var results = new ArrayList<TResult>(count);
            for (var i = 0; i < count; i++) {
                results.add(selector.apply(source.get(i)));
            }
            return results;
        }

        @Override
        public int getCount() {
            for (var item : source) {
                selector.apply(item);
            }
            return source.count();
        }

        @Override
        public AbstractIterator<TResult> skip(int count) {
            assert (count > 0);
            if (count >= source.count()) {
                return null;
            }
            return new ArraySkipTakeSelectIterator<>(source, selector, count, Integer.MAX_VALUE);
        }

        @Override
        public TResult tryGetFirst(boolean[] found) {
            assert (source.count() > 0);
            found[0] = true;
            return selector.apply(source.get(0));
        }
    }

    static final class EnumerableSelectIterator<TSource, TResult> extends AbstractIterator<TResult> {
        private final Enumerable<TSource> source;
        private final Function<TSource, TResult> selector;
        private Enumerator<TSource> enumerator;

        public EnumerableSelectIterator(Enumerable<TSource> source, Function<TSource, TResult> selector) {
            assert (source != null);
            assert (selector != null);
            this.source = source;
            this.selector = selector;
        }

        @Override
        protected AbstractIterator<TResult> getClone() {
            return new EnumerableSelectIterator<>(source, selector);
        }

        @Override
        public void close() {
            if (enumerator != null) {
                enumerator.close();
                enumerator = null;
            }
            super.close();
        }

        @Override
        public boolean moveNext() {
            switch (state) {
                case 1:
                    enumerator = source.enumerator();
                    state = 2;
                case 2:
                    assert (enumerator != null);
                    if (enumerator.moveNext()) {
                        current = selector.apply(enumerator.current());
                        return true;
                    }

                    close();
                    break;
            }

            return false;
        }


        @Override
        public <TResult2> Enumerable<TResult2> select(Function<TResult, TResult2> selector) {
            return new EnumerableSelectIterator<>(source, this.selector.andThen(selector));
        }

        @Override
        public List<TResult> toList() {
            var results = new ArrayList<TResult>();
            for (var item : this) {
                results.add(item);
            }
            return results;
        }

        @Override
        public int getCount() {
            int count = 0;

            for (var item : source) {
                selector.apply(item);
                count++;
            }

            return count;
        }

        @Override
        public TResult tryGetFirst(boolean[] found) {
            try (var e = source.enumerator()) {
                if (e.moveNext()) {
                    found[0] = true;
                    return selector.apply(e.current());
                }
                found[0] = false;
                return null;
            }
        }
    }

    static final class EnumerableIndexSelectIterator<TSource, TResult> implements Enumerator<TResult> {
        private final Enumerator<TSource> enumerator;
        private final BiFunction<TSource, Integer, TResult> selector;
        private int index = -1;
        private TResult current = null;

        public EnumerableIndexSelectIterator(Enumerator<TSource> enumerator, BiFunction<TSource, Integer, TResult> selector) {
            assert (enumerator != null);
            assert (selector != null);
            this.enumerator = enumerator;
            this.selector = selector;
        }


        @Override
        public void close() {
            enumerator.close();
            current = null;
        }

        @Override
        public boolean moveNext() {
            if (enumerator.moveNext()) {
                index++;
                current = selector.apply(enumerator.current(), index);
                return true;
            }

            close();
            return false;
        }

        @Override
        public TResult current() {
            return current;
        }
    }

    static final class ArraySkipTakeSelectIterator<TSource, TResult> extends AbstractIterator<TResult> {
        private final ArrayEnumerable<TSource> source;
        private final Function<TSource, TResult> selector;
        private final int minIndexInclusive;
        private final int maxIndexInclusive;

        public ArraySkipTakeSelectIterator(ArrayEnumerable<TSource> source, Function<TSource, TResult> selector, int minIndexInclusive, int maxIndexInclusive) {
            assert (source != null);
            assert (selector != null);
            assert (minIndexInclusive >= 0);
            assert (minIndexInclusive <= maxIndexInclusive);
            this.source = source;
            this.selector = selector;
            this.minIndexInclusive = minIndexInclusive;
            this.maxIndexInclusive = maxIndexInclusive;
        }

        @Override
        protected AbstractIterator<TResult> getClone() {
            return new ArraySkipTakeSelectIterator<>(source, selector, minIndexInclusive, maxIndexInclusive);
        }

        @Override
        public boolean moveNext() {
            var index = state - 1;
            if (index <= maxIndexInclusive - minIndexInclusive && index < source.count() - minIndexInclusive) {
                current = selector.apply(source.get(minIndexInclusive + index));
                ++state;
                return true;
            }

            close();
            return false;
        }

        @Override
        public <TResult2> Enumerable<TResult2> select(Function<TResult, TResult2> selector) {
            return new ArraySkipTakeSelectIterator<>(source, this.selector.andThen(selector), minIndexInclusive, maxIndexInclusive);
        }

        @Override
        public AbstractIterator<TResult> skip(int count) {
            assert (count > 0);
            var minIndex = minIndexInclusive + count;
            return minIndex > maxIndexInclusive ? null : new ArraySkipTakeSelectIterator<>(source, selector, minIndex, maxIndexInclusive);
        }

        @Override
        public List<TResult> toList() {
            int count = count();
            if (count == 0) {
                return new ArrayList<>();
            }

            var sourceIndex = minIndexInclusive;
            var list = new ArrayList<TResult>(count);
            for (var i = 0; i < count; i++, sourceIndex++) {
                list.add(selector.apply(source.get(sourceIndex)));
            }

            return list;
        }

        @Override
        public int getCount() {
            var count = count();

            int end = minIndexInclusive + count;
            for (int i = minIndexInclusive; i != end; ++i) {
                selector.apply(source.get(i));
            }

            return count;
        }
    }
}

File: src/main/kotlin/linq/Select.kt
package linq

import linq.collections.enumerable.ArrayEnumerable
import linq.collections.enumerable.EmptyEnumerable
import linq.collections.enumerable.SequenceEnumerable
import linq.exception.ExceptionArgument
import linq.exception.throwArgumentNullException
import kotlin.math.min

fun <TSource, TResult> select(source: Enumerable<TSource>?, selector: ((TSource) -> TResult)?): Enumerable<TResult> {
    if (source == null) {
        throwArgumentNullException(ExceptionArgument.Source)
    }
    if (selector == null) {
        throwArgumentNullException(ExceptionArgument.Selector)
    }

    if (source is AbstractIterator<TSource>) {
        return source.itSelect(selector)
    }

    if (source is ArrayEnumerable<TSource>) {
        if (source.count() == 0) return EmptyEnumerable()
        return ArraySelectIterator(source, selector)
    }

    return EnumerableSelectIterator(source, selector)
}

fun <TSource, TResult> select(
    source: Enumerable<TSource>?,
    selector: ((TSource, Int) -> TResult)?
): Enumerable<TResult> {
    if (source == null) {
        throwArgumentNullException(ExceptionArgument.Source)
    }

    if (selector == null) {
        throwArgumentNullException(ExceptionArgument.Selector)
    }

    if (isEmptyArray(source)) {
        return EmptyEnumerable()
    }

    return SequenceEnumerable(selectIterator(source, selector))
}

private fun <TSource, TResult> selectIterator(
    source: Enumerable<TSource>,
    selector: (TSource, Int) -> TResult
): Sequence<TResult> = sequence {
    var index = -1
    for (element in source) {
        index++
        yield(selector(element, index))
    }
}

internal class ArraySelectIterator<TSource, TResult>(
    private val source: ArrayEnumerable<TSource>,
    private val selector: (TSource) -> TResult
) : AbstractIterator<TResult>() {
    init {
        assert(source.size > 0)
    }

    override fun clone(): AbstractIterator<TResult> = ArraySelectIterator(source, selector)

    override fun moveNext(): Boolean {
        val index = state - 1
        if (index < source.size) {
            state++
            current = selector(source[index])
            return true
        }

        close()
        return false
    }

    override fun <TResult2> itSelect(selector: (TResult) -> TResult2): Enumerable<TResult2> =
        ArraySelectIterator(source, this.selector.then(selector))

    override fun getList(): List<TResult> {
        val count = source.size
        return List(count) { selector(source[it]) }
    }

    override fun itCount(): Int {
        for (item in source) {
            selector(item)
        }
        return source.size
    }

    override fun itSkip(count: Int): AbstractIterator<TResult>? {
        assert(count > 0)
        if (count >= source.size) {
            return null
        }
        return ArraySkipTakeSelectIterator(source, selector, count, Int.MAX_VALUE)
    }

    override fun itTake(count: Int): AbstractIterator<TResult>? {
        assert(count > 0)
        return if (count >= source.size) {
            this
        } else {
            ArraySkipTakeSelectIterator(source, selector, 0, count - 1)
        }
    }

    override fun tryGetElementAt(index: Int): TResult? {
        if (index.toUInt() < source.size.toUInt()) {
            return selector(source[index])
        }
        return null
    }

    override fun tryGetFirst(): TResult {
        assert(source.size > 0)
        return selector(source[0])
    }

    override fun tryGetLast(): TResult {
        assert(source.size > 0)
        return selector(source[source.size - 1])
    }

    override fun itContains(value: TResult): Boolean {
        for (element in source) {
            if (selector(element) == value) {
                return true
            }
        }

        return false
    }
}

internal class EnumerableSelectIterator<TSource, TResult>(
    private val source: Enumerable<TSource>,
    private val selector: (TSource) -> TResult
) : AbstractIterator<TResult>() {
    private var enumerator: Enumerator<TSource>? = null

    override fun clone(): AbstractIterator<TResult> {
        return EnumerableSelectIterator(source, selector)
    }

    override fun close() {
        enumerator?.close()
        enumerator = null
        super.close()
    }

    override fun moveNext(): Boolean {
        loop@ while (true) {
            when (state) {
                1 -> {
                    enumerator = source.enumerator()
                    state = 2
                    continue@loop
                }

                2 -> {
                    val enumerator = enumerator
                    enumerator.assertNotNull()
                    if (enumerator.moveNext()) {
                        current = selector(enumerator.current)
                        return true
                    }

                    close()
                    break@loop
                }
            }
        }
        return false
    }

    override fun <TResult2> itSelect(selector: (TResult) -> TResult2): Enumerable<TResult2> =
        EnumerableSelectIterator(source, this.selector.then(selector))

    override fun getList(): List<TResult> {
        return source.mapTo(arrayListOf(), selector)
    }

    override fun itCount(): Int {
        var count = 0

        for (item in source) {
            selector(item)
            count++
        }

        return count
    }

    override fun tryGetElementAt(index: Int): TResult? {
        var index = index
        if (index >= 0) {
            source.enumerator().use {
                while (it.moveNext()) {
                    if (index == 0) {
                        return selector(it.current)
                    }

                    index--
                }
            }
        }

        return null
    }

    override fun tryGetFirst(): TResult? {
        source.enumerator().use {
            if (it.moveNext()) {
                return selector(it.current)
            }
            return null
        }
    }

    override fun tryGetLast(): TResult? {
        source.enumerator().use {
            if (it.moveNext()) {
                var last = it.current

                while (it.moveNext()) {
                    last = it.current
                }

                return selector(last)
            }
        }

        return null
    }
}

class ArraySkipTakeSelectIterator<TSource, TResult>(
    private val source: ArrayEnumerable<TSource>,
    private val selector: (TSource) -> TResult,
    private val minIndexInclusive: Int,
    private val maxIndexInclusive: Int
) : AbstractIterator<TResult>() {
    init {
        assert(minIndexInclusive >= 0)
        assert(minIndexInclusive <= maxIndexInclusive)
    }

    override fun clone(): AbstractIterator<TResult> {
        return ArraySkipTakeSelectIterator(source, selector, minIndexInclusive, maxIndexInclusive)
    }

    override fun moveNext(): Boolean {
        val index = state - 1
        if (index.toUInt() <= (maxIndexInclusive - minIndexInclusive).toUInt() &&
            index < source.size - minIndexInclusive
        ) {
            current = selector(source[minIndexInclusive + index])
            ++state
            return true
        }

        close()
        return false
    }

    override fun <TResult2> itSelect(selector: (TResult) -> TResult2): Enumerable<TResult2> =
        ArraySkipTakeSelectIterator(
            source,
            this.selector.then(selector),
            minIndexInclusive,
            maxIndexInclusive
        )

    override fun itSkip(count: Int): AbstractIterator<TResult>? {
        assert(count > 0)
        val minIndex = minIndexInclusive + count
        return if (minIndex.toUInt() > maxIndexInclusive.toUInt()) {
            null
        } else {
            ArraySkipTakeSelectIterator(source, selector, minIndex, maxIndexInclusive)
        }
    }

    override fun itTake(count: Int): AbstractIterator<TResult>? {
        assert(count > 0)
        val maxIndex = minIndexInclusive + count - 1
        return if (maxIndex.toUInt() >= maxIndexInclusive.toUInt()) {
            this
        } else {
            ArraySkipTakeSelectIterator(source, selector, minIndexInclusive, maxIndex)
        }
    }

    override fun tryGetElementAt(index: Int): TResult? {
        if (index.toUInt() <= (maxIndexInclusive - minIndexInclusive).toUInt() &&
            index < source.size - minIndexInclusive
        ) {
            return selector(source[minIndexInclusive + index])
        }

        return null
    }

    override fun tryGetFirst(): TResult? {
        if (source.size > minIndexInclusive) {
            return selector(source[minIndexInclusive])
        }
        return null
    }

    override fun tryGetLast(): TResult? {
        val lastIndex = source.size - 1
        if (lastIndex >= minIndexInclusive) {
            return selector(source[min(lastIndex, maxIndexInclusive)])
        }

        return null
    }

    val count: Int
        get() {
            val count = source.size
            if (count <= minIndexInclusive) {
                return 0
            }
            return min(count - 1, maxIndexInclusive) - minIndexInclusive + 1
        }

    override fun getList(): List<TResult> {
        val count = count
        if (count == 0) {
            return arrayListOf()
        }

        var sourceIndex = minIndexInclusive
        val list = arrayListOf<TResult>()
        repeat(count) {
            list.add(selector(source[sourceIndex]))
            sourceIndex++
        }

        return list
    }

    override fun itCount(): Int {
        val count = count

        val end = minIndexInclusive + count
        for (i in minIndexInclusive until end) {
            selector(source[i])
        }

        return count
    }

    override fun itContains(value: TResult): Boolean {
        val count = count

        val end = minIndexInclusive + count

        return (minIndexInclusive until end)
            .any { selector(source[it]) == value }
    }
}